lib/calendar.ts

```diff
@@ -455,9 +455,213 @@ impl['iso8601'] = {
+// to validate the Temporal API and to get feedback. However, non-ISO calendar
+// implementation is subject to change because these calendars are
+// implementation-defined.
+//
+// Some ES implementations don't include ECMA 402. For this reason, it's helpful
+// to ensure a clean separation between the ISO calendar implementation which is
+// a part of ECMA 262 and the non-ISO calendar implementation which requires
+// ECMA 402.
+//
+// To ensure this separation, the implementation is split. The `NonIsoImpl`
+// interface is the top-level implementation for all non-ISO calendars. This
+// type has the same shape as the ECMA 262-only ISO calendar implementation so
+// can use the same callers, tests, etc.
+//
+// A derived interface `NonIsoImplWithHelper` adds a `helper` property that
+// includes the remaining non-ISO implementation properties and methods beyond
+// the ISO implementation above. The `helper` property's shape is a base
+// singleton object common to all calendars (`HelperSharedImpl`) that's extended
+// (interface `HelperPerCalendarImpl`) with implementation that varies for each
+// calendar.
+//
+// Typing of individual methods in the interfaces below uses the `this`
+// "parameter" declaration definition, which is a fake parameter (stripped by TS
+// during compilation and not visible at runtime) that tells TS what type `this`
+// is for a method. For historical reasons, the initial implementation of
+// non-ISO calendars mirrored the code style of the previous ISO-only
+// implementation which didn't use ES6 classes. Using the `this` parameter is a
+// hack to delay converting this file to use ES6 classes until the code was
+// fully typed to make a `class` refactoring easier and safer. We'll probably do
+// this conversion in the future. (PRs welcome!)
+
+/**
+ * `NonIsoImpl` - The generic top-level implementation for all non-ISO
+ * calendars. This type has the same shape as the 262-only ISO calendar
+ * implementation, which means the `Calendar` class implementation can swap out
+ * the ISO for non-ISO implementations without changing any `Calendar` code.
+ */
+interface NonIsoImpl {
+  dateFromFields(
+    this: NonIsoImplWithHelper,
+    fieldsParam: Params['dateFromFields'][0],
+    options: NonNullable<Params['dateFromFields'][1]>,
+    calendar: Temporal.Calendar
+  ): Temporal.PlainDate;
+  yearMonthFromFields(
+    this: NonIsoImplWithHelper,
+    fieldsParam: Params['yearMonthFromFields'][0],
+    options: NonNullable<Params['yearMonthFromFields'][1]>,
+    calendar: Temporal.Calendar
+  ): Temporal.PlainYearMonth;
+  monthDayFromFields(
+    this: NonIsoImplWithHelper,
+    fieldsParam: Params['monthDayFromFields'][0],
+    options: NonNullable<Params['monthDayFromFields'][1]>,
+    calendar: Temporal.Calendar
+  ): Temporal.PlainMonthDay;
+  fields(fieldsParam: string[]): Return['fields'];
+  mergeFields(fields: Params['mergeFields'][0], additionalFields: Params['mergeFields'][1]): Return['mergeFields'];
```

Comment:
No idea if this would work, but is it possible to do something like `mergeFields: typeof Temporal.Calendar.prototype.mergeFields`? Just out of curiosity.

Reply:
Is this something that will go away when we use ES6 classes here? I ask because I find these types a bit difficult to read...

Reply:
Here's what the equivalent will look like in ES6 classes: https://github.com/js-temporal/temporal-polyfill/pull/113/files#diff-0751307e1b8622b923dfb7f18f7aed915f397cd1f6a67a5465d9672db48839f5R478

Is that better?

Reply:
I have a vague hunch that this is more complicated than it needs to be, but I'm not prepared to spend time to prove it, so I won't complain anymore :smile:

Reply:
"I have a vague hunch that this is more complicated than it needs to be" could be the motto of half of my time on this planet, I think. ;-)